File: ScalaKurs(Knabe)_J/src/ueb02/ListTest.java
package ueb02;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class ListTest {

	private static final List _empty = null;
	private static final List _list = new List(1, new List(2, new List(3, null)));

	@Test
	public void prependValueToEmptyList() {
		final List actual = prepend(9, _empty);
		assertEquals(new List(9, null), actual);
		assertEquals("9", actual.toString());
	}

	@Test
	public void attachValueToEmptyList() {
		final List actual = attach(9, null);
		assertEquals(new List(9, null), actual);
		assertEquals("9", actual.toString());
	}

	@Test
	public void prependValueToListWithElements() {
		final List actual = prepend(9, _list);
		assertEquals(new List(9, _list), actual);
		assertEquals("9 1 2 3", actual.toString());
	}
	@Test
	public void attachValueToListWithElements() {
		final List actual = attach(9, _list);
		assertEquals(new List(1, new List(2, new List(3, new List(9, null)))), actual);
		assertEquals("1 2 3 9", actual.toString());
	}

	/**
	 * Prepends the given value to the fron of the given list.
	 */
	private List prepend(final int value, final List list) {
		return new List(value, list);
	}

	/**
	 * Attaches the given value to the end of the given list.
	 */
	private List attach(final int value, final List list) {
		return list == null 
				? new List(value, null) 
				: prepend(list.head, attach(value, list.tail));
	}
}

File: ScalaKurs(Knabe)_S/src/ueb02/ListTest.scala
package ueb02

import org.junit.Assert.assertEquals
import org.junit.Test

class ListTest {

  val _empty = null
  val _list = List(1, List(2, List(3, null)))

  @Test
  def prependValueToEmptyList() {
    val actual = prepend(9, _empty);
    assertEquals("9", actual.toString)
  }

  @Test 
  def attachValueToEmptyList() {
    val actual = attach(9, _empty);
    assertEquals("9", actual.toString)
  }

  @Test 
  def prependValueToListWithElements() {
    val actual = prepend(9, _list);
    assertEquals("9 1 2 3", actual.toString)
  }

  @Test 
  def attachValueToListWithElements() {
    val actual = attach(9, _list);
    assertEquals("1 2 3 9", actual.toString)
  }

  /**
   * Prepends the given value to the fron of the given list.
   */
  def prepend(value: Int, list: List): List = {
    new List(value, list)
  }

  /**
   * Attaches the given value to the end of the given list.
   */
  def attach(value: Int, list: List): List = list match {
    case null => List(value, null)
    case _    => prepend(list.head, attach(value, list.tail))
  }
}
